gobblin-modules/gobblin-kafka-common/src/main/java/org/apache/gobblin/source/extractor/extract/kafka/validator/TopicValidators.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.gobblin.source.extractor.extract.kafka.validator;
+
+import com.google.common.base.Strings;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.apache.gobblin.configuration.SourceState;
+import org.apache.gobblin.source.extractor.extract.kafka.KafkaTopic;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The TopicValidators contains a list of {@link TopicValidatorBase} that validate topics.
+ * To enable it, add below settings in the config:
+ *   gobblin.kafka.topicValidators=validator1_class_name;validator2_class_name...
+ */
+public class TopicValidators {
+  public static final String VALIDATOR_CLASSES_KEY = "gobblin.kafka.topicValidators";
+
+  public static final String VALIDATOR_CLASS_DELIMITER = ";";
+  private static final Logger LOGGER = LoggerFactory.getLogger(TopicValidators.class);
+
+  private final List<TopicValidatorBase> validators = new ArrayList<>();
+
+  public TopicValidators(SourceState state) {
+    String validatorClasses = state.getProp(VALIDATOR_CLASSES_KEY);
+    if (Strings.isNullOrEmpty(validatorClasses)) {
+      return;
+    }
+
+    String[] validatorClassNames = validatorClasses.split(VALIDATOR_CLASS_DELIMITER);
+    Arrays.stream(validatorClassNames).forEach(validator -> {
+      try {
+        this.validators.add(
+            (TopicValidatorBase) Class.forName(validator).getConstructor(SourceState.class).newInstance(state));
+      } catch (Exception e) {
+        LOGGER.error("Failed to create topic validator: {}, due to {}", validator, e);
+      }
+    });
+  }
+
+  /**
+   * Validate topics with all the internal validators.
+   * Note: the validations for every topic run in parallel.
+   * @param topics the topics to be validated
+   * @return the topics that pass all the validators
+   */
+  public List<KafkaTopic> validate(List<KafkaTopic> topics) {
+    // Validate the topics in parallel
+    return topics.parallelStream()
```

Comment:
Not that I think this class will be used within Incremental Compute, but just FYI, we have seen in the past where parallelStream has caused some strange bugs related to not picking up dependencies in Spark. In general, I'd suggest always thinking about whether the parallelism is important given that the api has caused issue in the past.

https://github.com/apache/gobblin/pull/3706

Reply:
Thanks for the reference of that PR.

Yes, the parallelism is on purpose. Some validator could be slow, validating the topics sequentially would significantly slow down the startup.

For example, there will be [another PR to add an OrcSchemaConversionValidator](https://github.com/apache/gobblin/pull/3794/files#diff-c480e84c1e7ab69ed31e521c5615de4b270ce6da72333a29ec779b47f04e3d0aR45-R48) that queries the remote schema registry for every topic.

If the `parallelStream` is not suggested, we would have to do it in a separate thread pool while trying to avoid the context class loader issue. Any comment on this approach?

Reply:
Do we intend to keep this code limited to Kafka source? i.e. do we see a world where we'd want to use this topic validation in incremental?..

If we do see spark uses cases, could we look into the following comment from that PR that describes an alternative approach that would set the appropriate class loader? This sort of topic validation could make sense since Iceberg does not support some of our weird edge cases (e.g. non optional union schema)

https://github.com/apache/gobblin/pull/3706#issuecomment-1596241791

JFYI, in compaction we do have a similar validation step at the dataset finder level (and the PR I linked previously is a bug fix for that feature). If the intention is to use this topic validation across the board you can consider whether it makes sense to converge these code paths.  
- https://github.com/apache/gobblin/pull/3632
- https://github.com/apache/gobblin/pull/3648

Reply:
Ideally, the data from a bad topic should not come to the incremental compaction stage, because:
1. With static validators, `KafkaSource` can skip the bad topics during the work unit creation.
2. With dynamic validators, bad topics are detected on the fly and then FastIngest can be restarted to skip those bad topics.

However, the incremental compaction may potentially use a different set of validators, so it still make sense to provide the validation support beyond FastIngest. So let's go for it.

To set the context class loader correctly, in the new iteration the `ExecutorsUtils` is enhanced to help create thread pools where the running tasks can have the same access control and class loader settings as the thread that submits the tasks. Please help take a look.